cupyx/scipy/sparse/linalg/_iterative.py

```diff
@@ -202,6 +202,103 @@ def gmres(A, b, x0=None, tol=1e-5, restart=None, maxiter=None, M=None,
     return mx, info
 
 
+def cgs(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None,
+        atol=None):
+    """Use Conjugate Gradient Squared iteration to solve ``Ax = b``.
+
+    Args:
+        A (ndarray, spmatrix or LinearOperator): The real or complex matrix of
+            the linear system with shape ``(n, n)``.
+        b (cupy.ndarray): Right hand side of the linear system with shape
+            ``(n,)`` or ``(n, 1)``.
+        x0 (cupy.ndarray): Starting guess for the solution.
+        tol (float): Tolerance for convergence.
+        maxiter (int): Maximum number of iterations.
+        M (ndarray, spmatrix or LinearOperator): Preconditioner for ``A``.
+            The preconditioner should approximate the inverse of ``A``.
+            ``M`` must be :class:`cupy.ndarray`,
+            :class:`cupyx.scipy.sparse.spmatrix` or
+            :class:`cupyx.scipy.sparse.linalg.LinearOperator`.
+        callback (function): User-specified function to call after each
+            iteration. It is called as ``callback(xk)``, where ``xk`` is the
+            current solution vector.
+        atol (float): Tolerance for convergence.
+
+    Returns:
+        tuple:
+            It returns ``x`` (cupy.ndarray) and ``info`` (int) where ``x`` is
+            the converged solution and ``info`` provides convergence
+            information.
+
+    .. seealso:: :func:`scipy.sparse.linalg.cgs`
+    """
+    A, M, x, b = _make_system(A, M, x0, b)
+
+    matvec = A.matvec
+    psolve = M.matvec
+
+    n = A.shape[0]
+    if n == 0:
+        return cupy.empty_like(b), 0
+    b_norm = cupy.linalg.norm(b)
+    if b_norm == 0:
+        return b, 0
+    if atol is None:
+        atol = tol * float(b_norm)
+    else:
+        atol = max(float(atol), tol * float(b_norm))
+    if maxiter is None:
+        maxiter = n * 5
+
+    r0 = b - matvec(x)
+
+    rho = cupy.dot(r0, r0)
+    r_norm = numpy.sqrt(rho)
```

Comment:
Please use `cupy.sqrt` to take a `cupy.ndarray`.
```suggestion
    r_norm = cupy.sqrt(rho)
```

Reply:
I used numpy as r_norm has only a single value. And doing sqrt of single value on numpy will be faster than on cupy.

Reply:
Okay, I got that you want `r_norm` to be a numpy value. However, `rho` is the return value of `cupy.dot`, which is a `cupy.ndarray`, and `numpy.sqrt` taking a `cupy.ndarray` dispatches to `cupy.sqrt`, eventually returning a `cupy.ndarray` for `r_norm`. Please check `print(type(r_norm))`. The following is what you should do.
```suggestion
    r_norm = cupy.sqrt(rho).get().item()
```

FYI:
```
$ python -c 'import numpy as np, cupy as cp; print(type(np.sqrt(cp.array(4))))'
<class 'cupy._core.core.ndarray'>
```

Reply:
BTW, do we need to compute `r_norm` here?

Reply:
> numpy.sqrt taking a cupy.ndarray dispatches to cupy.sqrt

I was unaware of this. 
And we don't need to calculate r_norm there, I did that for some check while writing it. So will remove this part.